src/kill-parallel.c

```diff
@@ -0,0 +1,261 @@
+
+  int ps_n = length(ps);
+
+  SEXP killed_shelter = PROTECT(allocVector(LGLSXP, ps_n));
+  int* killed = LOGICAL(killed_shelter);
+
+  int killed_n = 0;
+  memset(killed, 0, sizeof(int) * ps_n);
+
+  int fds[2] = { -1 };
+  if (pipe(fds)) {
+    ps__set_error("Can't create pipe for parallel kill.");
+    err = ps__last_error;
+    goto cleanup;
+  }
+  kill_parallel_pipe = fds[1];
+
+  ps__nonblock_fcntl(fds[0], 1);
+  ps__nonblock_fcntl(fds[1], 1);
+
+  for (int i = 0; i < ps_n; ++i) {
+    ps_handle_t *handle = R_ExternalPtrAddr(VECTOR_ELT(ps, i));
+    if (psll__is_running(handle)) {
+      kill(handle->pid, SIGTERM);
+    }
+  }
+
+  struct pollfd fd = {
+    .fd = fds[0],
+    .events = POLLIN,
+    .revents = 0
+  };
+  unblock_sigchld();
+
+  int timeleft = ceil(REAL(ffi_grace)[0] * 1000);
+  if (timeleft < 0) {
+    ps__set_error("`grace` can't be negative.");
+    err = ps__last_error;
+    goto cleanup;
+  }
+
+  int ready = 0;
+
+  while (timeleft > PS_INTERRUPT_INTERVAL) {
+    do {
+      ready = poll(&fd, 1, PS_INTERRUPT_INTERVAL);
+    } while (ready == -1 && errno == EINTR);
+
+    // Rough ETA computation that doesn't take into account time spent
+    // until woken up by a SIGCHLD
+    if (ready == 0) {
+      timeleft -= PS_INTERRUPT_INTERVAL;
+    }
+
+    char buf[1];
+    int res = 0;
+
+    do {
+      res = read(fds[0], buf, 1);
+    } while (res == 1 || (res == -1 && errno == EINTR));
+
+    // Should we restore our sigchld handler in case it was overridden
+    // by R code?
+    if ((err = r_check_for_user_interrupt(cont))) {
+      goto cleanup;
+    }
+
+    for (int i = 0; i < ps_n; ++i) {
+      if (killed[i]) {
+        continue;
+      }
+
+      ps_handle_t *handle = R_ExternalPtrAddr(VECTOR_ELT(ps, i));
+
+      if (!psll__is_running(handle)) {
+        killed_n += 1;
+        killed[i] = 1;
+      }
+    }
+
+    if (killed_n == ps_n) {
+      goto cleanup;
+    }
+  }
+
+  if (killed_n == ps_n) {
+    goto cleanup;
+  }
+
+  /* Maybe we are not done, and there is a little left from the timeout */
+  if (ready == 0 && timeleft >= 0) {
+    do {
+      ready = poll(&fd, 1, timeleft);
+    } while (ready == -1 && errno == EINTR);
+  }
+
+  // If some processes are still running after the grace period,
+  // resort to SIGKILL
+  for (int i = 0; i < ps_n; ++i) {
+    if (killed[i]) {
+      continue;
+    }
+
+    ps_handle_t *handle = R_ExternalPtrAddr(VECTOR_ELT(ps, i));
+
+    if (psll__is_running(handle)) {
+      kill(handle->pid, SIGKILL);
+    }
+  }
+
+ cleanup:
+  kill_parallel_pipe = -1;
+
+  if (fds[0] >= 0) close(fds[0]);
+  if (fds[1] >= 0) close(fds[1]);
+
+  kill_parallel_remove_sigchld(&old_action);
+  procmask_set(&old_mask);
+
+  if (err) {
+    r_unwind(cont);
+  }
+
+  UNPROTECT(2);
+  return R_NilValue;
+}
+
+
+// SIGCHLD -------------------------------------------------------------
+
+static
+void kill_parallel_sigchld_callback(int sig, siginfo_t *info, void *ctx) {
```

Comment:
The parallel kill routine listens for termination events via this `SIGCHLD` handler. It notifies our kill routine via a pipe as well as the previous chld handler so they can collect terminated processes.